src/constructors/lattices.c

```diff
@@ -0,0 +1,308 @@
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_rectangle_shape(igraph_t *graph, igraph_integer_t size_x,
+                                                       igraph_integer_t size_y, igraph_bool_t directed, igraph_bool_t mutual)
+{
+    igraph_integer_t row_count = size_y;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_CHECK(igraph_vector_int_init(&row_lengths_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_lengths_vector);
+    IGRAPH_CHECK(igraph_vector_int_init(&row_start_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_start_vector);
+
+    for (i = 0; i < row_count; i++)
+    {
+        VECTOR(row_lengths_vector)[i] = size_x;
+        VECTOR(row_start_vector)[i] = (row_count - i) / 2;
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_hex_shape(igraph_t *graph, igraph_integer_t size_x,
+                                                 igraph_integer_t size_y, igraph_integer_t size_z, igraph_bool_t directed, igraph_bool_t mutual)
+{
+    igraph_integer_t row_count = size_y + size_z - 1;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_CHECK(igraph_vector_int_init(&row_lengths_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_lengths_vector);
+    IGRAPH_CHECK(igraph_vector_int_init(&row_start_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_start_vector);
+
+    igraph_integer_t row_length = size_x;
+    igraph_integer_t row_start = size_y - 1;
+    igraph_integer_t first_threshold = fmin(size_y - 1, size_z - 1);
+    igraph_integer_t second_threshold = fmax(size_y - 1, size_z - 1);
+    igraph_integer_t sgn_flag = size_y < size_z ? 0 : -1;
+
+    for (i = 0; i < row_count; i++)
+    {
+        VECTOR(row_lengths_vector)[i] = row_length;
+        VECTOR(row_start_vector)[i] = row_start;
+
+        if (i < first_threshold)
+        {
+            row_length++;
+            row_start--;
+        }
+        else if (i < second_threshold)
+        {
+            row_start += sgn_flag;
+        }
+        else
+        {
+            row_length--;
+        }
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(1);
```

Comment:
The first `IGRAPH_FINALLY_CLEAN(...)` is not necessary and the second one should be `IGRAPH_FINALLY_CLEAN(2)`.